### PR TITLE
[NPM] Use single connection buffer for getting connections

### PR DESCRIPTION
**Pull request by hmahmood**
### What does this PR do?

Uses a single `ConnectionBuffer` object to accumulate and report connections during a call to `Tracer.GetActiveConnections()`. Previously we were using one `ConnectionBuffer` object to collect active connections from ebpf and another to compute the delta of connections' stats, to which the connections from the first buffer were copied to. Even with one `ConnectionBuffer` allocations are avoided till the very end of the delta computation function (networkState.GetDelta). This is done by reusing the slices passed in to `GetDelta`.

Some changes to aggregation also:
* a list of connections instead of one connection per aggregation key is now maintained. This is necessary when we can't aggregate connections' IP translation field. Each element in the list of connections per aggregation differs only in IP translation
* closed and active connections are aggregated with each other. This should be ok to do since we sum together the stats, and will reduce connection count reported
* the `ProtocolStack` field was not being aggregated at all

Gains from internal testing deployments below. Max RSS seems to show more improvement, while the max for working set seems to be lower as well, which should help with spikes in mem usage.

k8s RSS memory

k8s working set memory

### Reviewer's Checklist

- [x] If known, an appropriate milestone has been selected; otherwise the `Triage` milestone is set.
- [ ] Use the `major_change` label if your change either has a major impact on the code base, is impacting multiple teams or is changing important well-established internals of the Agent. This label will be use during QA to make sure each team pay extra attention to the changed behavior. For any customer facing change use a releasenote.
- [x] A [release note](https://github.com/DataDog/datadog-agent/blob/main/docs/dev/contributing.md#reno) has been added or the `changelog/no-changelog` label has been applied.
- [x] Changed code has automated tests for its functionality.
- [x] Adequate QA/testing plan information is provided if the `qa/skip-qa` label is not applied.
- [x] At least one `team/..` label has been applied, indicating the team(s) that should QA this change.
- [ ] If applicable, docs team has been notified or [an issue has been opened on the documentation repo](https://github.com/DataDog/documentation/issues/new).
- [ ] If applicable, the `need-change/operator` and `need-change/helm` labels have been applied.
- [ ] If applicable, the `k8s/<min-version>` label, indicating the lowest Kubernetes version compatible with this feature.
- [ ] If applicable, the [config template](https://github.com/DataDog/datadog-agent/blob/main/pkg/config/config_template.yaml) has been updated.

### CODE REVIEW

**Review by brycekahle — APPROVED** [review]
only reviewed files owned by ebpf-platform

File: pkg/network/state.go [hunk 1]
```diff
@@ -241,6 +231,20 @@ func (ns *networkState) GetTelemetryDelta(
...
+		// swap first so that the connection pointer
+		// passed to keep will be stable
+		conns[p], conns[i] = conns[i], conns[p]
+		if keep(&conns[p]) {
+			p++
+		}
+	}
+
+	return conns[:p]
+}
```
Comment (guyarb) [inline pkg/network/state.go:246]: Please add UTs for that function

**Review by guyarb — APPROVED** [review]

File: pkg/network/slice/chain.go [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+func (c Chain[T]) Get(i int) T {
+	for _, s := range c.slices {
+		if i < len(s) {
+			return s[i]
+		}
+
+		i -= len(s)
+	}
+
+	panic("index out of range")
```
[the remark this answers is not in the document]
Comment (hmahmood) [reply pkg/network/slice/chain.go:53]: Bad index for an array/slice is always a panic, and I think it should be a panic here too. Error just would seem very out of place for this function, and for callers to handle.

File: tools/ebpf/Dockerfiles/Dockerfile-process-agent-dev [hunk 3]
```diff
@@ -3,7 +3,7 @@ FROM $AGENT_BASE
 ARG CORE_AGENT_DEST
 
 # include some useful dev tools since this will be used or development
-RUN apt-get update -y && apt-get install -y jq conntrack netcat dnsutils iproute2 net-tools
+RUN apt-get update -y && apt-get install -y jq conntrack netcat-openbsd dnsutils iproute2 net-tools
```
[the remark this answers is not in the document]
Comment (hmahmood) [reply tools/ebpf/Dockerfiles/Dockerfile-process-agent-dev:6]: Great, I have another one: https://github.com/DataDog/datadog-agent/pull/18900 :-)

**Review by derekwbrown — COMMENTED** [review]
Has this been installed & run under any sort of load on Windows?

**Comment by hmahmood** [comment]
> Has this been installed & run under any sort of load on Windows?

Not on windows, no

**Review by derekwbrown — APPROVED** [review]

**Pull request merged**